airbyte-featureflag/src/main/kotlin/Client.kt

```diff
@@ -0,0 +1,196 @@
+import com.fasterxml.jackson.module.kotlin.readValue
+import com.fasterxml.jackson.module.kotlin.registerKotlinModule
+import com.launchdarkly.sdk.ContextKind
+import com.launchdarkly.sdk.LDContext
+import com.launchdarkly.sdk.LDUser
+import com.launchdarkly.sdk.server.LDClient
+import java.lang.Thread.MIN_PRIORITY
+import java.nio.file.Path
+import java.nio.file.StandardWatchEventKinds
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.read
+import kotlin.concurrent.thread
+import kotlin.concurrent.write
+import kotlin.io.path.isRegularFile
+
+/**
+ * Open Questions:
+ * 1. Can the context and flag be combined into one class?
+ *   - i.e. Does it make sense to one a feature-flag apply to two different context types?
+ * 2. Should the default value be merged into the Flag type instead of specified on each check?
+ *   - Or maybe still allow it to be specified, but make it an optional parameter
+ */
+
+/**
+ * Feature-Flag Client interface.
+ */
+sealed interface FeatureFlagClient {
+    /**
+     * Returns true if the flag with the provided context should be enabled. Returns false otherwise.
+     */
+    fun enabled(flag: Flag, ctx: Context): Boolean
+}
+
+/**
+ * Platform feature-flag client. Feature-flag are derived from a yaml config file.
+ * Also supports flags defined via environment-variables via the [EnvVar] class.
+ *
+ * @param [config] the location of the yaml config file that contains the feature-flag definitions.
+ * The [config] will be watched for changes and the internal representation of the [config] will be updated to match.
+ */
+class PlatformClient(config: Path) : FeatureFlagClient {
+    /** [flags] holds the mappings of the flag-name to the flag properties */
+    private var flags: Map<String, PlatformFlag> = readConfig(config)
+
+    /** lock is used for ensuring access to the flags map is handled correctly when the map is being updated. */
+    private val lock = ReentrantReadWriteLock()
+
+    init {
+        if (!config.isRegularFile()) {
+            throw IllegalArgumentException("config must reference a file")
+        }
+
+        config.onChange {
+            val tempFlags = readConfig(config)
+            lock.write { flags = tempFlags }
+        }
+    }
+
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
+        return when (flag) {
+            is EnvVar -> flag.enabled()
+            else -> lock.read { flags[flag.key]?.enabled ?: flag.default }
+        }
+    }
+}
+
+/**
+ * Cloud feature-flag client. Feature-flags are derived from an external source (the LDClient).
+ * Also supports flags defined via environment-variables via the [EnvVar] class.
+ *
+ * @param [client] the Launch-Darkly client for interfacing with Launch-Darkly.
+ */
+class CloudClient(private val client: LDClient) : FeatureFlagClient {
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
+        return when (flag) {
+            is EnvVar -> flag.enabled()
```

Comment:
The `EnvVar` flags will ideally be a temporary stop gap on the migration to a config-file/LaunchDarkly and I feel that the added complexity of a `CompositeClient` probably isn't worth it.  Additionally during the migration to using new feature-flags there will be times when both the environment-variable and the externally-controlled feature-flag would both be defined, but we only would want one source of truth for these and probably on a flag-by-flag basis.

Originally I wasn't planning on having `EnvVar` flag support in all the clients, but ultimately realized it was necessary.  It might be worth looking into changing the `sealed interface FeatureFlagClient` into a `sealed class FeatureFlagClient` with a method on it that could dedupe some of this.

In my ideal world, the `EnvVar` code is removed within weeks of us moving over to LaunchDarkly.